ec/ecresource/deploymentresource/plan_modifier.go

```diff
@@ -0,0 +1,58 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package deploymentresource
+
+import (
+	"context"
+	"github.com/elastic/cloud-sdk-go/pkg/api/deploymentapi/deptemplateapi"
+	deploymentv2 "github.com/elastic/terraform-provider-ec/ec/ecresource/deploymentresource/deployment/v2"
+	"github.com/elastic/terraform-provider-ec/ec/ecresource/deploymentresource/planmodifiers"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+)
+
+var _ resource.ResourceWithModifyPlan = &Resource{}
+
+func (r Resource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	if req.Plan.Raw.IsNull() {
+		// Resource is being destroyed
+		return
+	}
+
+	var plan deploymentv2.DeploymentTF
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	template, err := deptemplateapi.Get(deptemplateapi.GetParams{
```

Comment:
Also, does it make sense to check whether `master` tier isn't explicitly configured before making any API calls?

Reply:
Yeah makes sense.

Regarding the template migration: I have to test it, as it isn't clear how it behaves. I think it could lead to problems if the default size for the master tier changes as that update probably wouldn't be applied.

Reply:
Tested migrating the template, and that works already since the modifier loads the new template and uses the values from there. 
I think the only case that could fail is where the template stays the same, but only the IC changes. Still have to test that case.

Reply:
Do we have to test the case when `migrate_to_latest_hardware` is specified? Does the plan contains the new DT in this case?

Reply:
Yeah that I still have to test. I think `migrate_to_latest_hardware` is supposed to keep the same DT, but change to the latest ICs.

Reply:
`migrate_to_latest_hardware` works like this:
- By default, every plan just re-uses the values from state (template if there is no value yet, for create)
- `migrate_to_latest_hardware` will reset the value and therefore get the latest value from the template again
- We don't need to look at the `migrateTemplateRequest` as it just contains the full request to update the deployment. But for the dedicated master we anyway only look at the template, which is already loaded separately.

So basically there is currently one problem:
- The modifier will update to the latest IC version even when `migrate_to_latest_hardware` is not set.
- So the modifier should not recalculate the size/zone_count on every plan
- It should only update in the following cases:
  - Node-count has changed and tier should be enabled -> In this case it is fine to use the latest IC
  - `migrate_to_latest_hardware` is set -> Also use the default from the latest IC

Reply:
> Node-count has changed and tier should be enabled -> In this case it is fine to use the latest IC
Why? (if `migrate_to_latest_hardware` isn't set?)

So now IIUC, the same template can return the new IC version with other values for `DiscreteSizes` and `MaxZones`. But IMO the dedicated `master` tier plan modifier should use the old values for these fields if `migrate_to_latest_hardware` isn't set.

If this is correct, does it make sense to store`DiscreteSizes` and `MaxZones` in state how we do currently for `InstanceConfigurationId`?

Reply:
Though I tested how it works in the UI, and the UI does always keep the same version. So maybe it is better to not go to the latest IC when enabling the master.

Reply:
IIUC the decision to switch to the latest IC should be taken by the user (because it can incur additional costs).
I guess this is one of the reasons behind introducing `migrate_to_latest_hardware` attribute.

> I was originally thinking about using state, but it has some drawbacks: It will also show up in the plan (though the private state could be used for that). And makes everything more complicated because the state is only available in the update-case. So all logic needs to try state, if not available fall back to template.

We can make it read-only and use state for unknown or similar plan modifier (similar to `latest_instance_configuration_id`). I think this is a pretty TF native approach though loading the requested IC version can be a valid alternative.

Reply:
I like the idea of just loading the right IC when needed more. Seems simpler to do one request when needed.

Adding a new field to the state also means having to keep it up-to-date in the read-step for all tiers. So it would still mean having to read it, though it could be skipped if the value is already in the state. And for this specific field I'm not sure if it is very useful for users to see as it is not really the default for the tier (that is set in the tier of the deployment template instead).

Reply:
Though I just realized this IC API only works in ECE 🤔 As it turns out the right way to get them is as part of the deployments API, with the parameter `show_instance_configurations=true`.

Reply:
Another SDK change is needed to be able to get the ICs when loading the deployment: https://github.com/elastic/cloud-sdk-go/pull/469

Reply:
The code currently doesn't compile as it needs another SDK update, but otherwise the PR should now cover everything.

I am now loading the ICs as part of the deployment-read (and storing them in private state), falling back to the template if that isn't available (for create for example).